File: src/com/easy/LongestUnivaluePath_687.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class LongestUnivaluePath_687 {

    // Use recursive:
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 4 ms, faster than 65.47%
    // Memory Usage: 40.6 MB, less than 98.87%
    private int mAnswer = 0;

    public int longestUnivaluePath(TreeNode root) {
        if (root == null) {
            return 0;
        }
        univaluePath(root);

        return mAnswer;
    }

    public int univaluePath(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int mLeft = univaluePath(root.left);
        int mRight = univaluePath(root.right);
        int mPathLeft = 0, mPathRight = 0;
        if (root.left != null && root.val == root.left.val) {
            mPathLeft = mLeft + 1;
        }
        if (root.right != null && root.val == root.right.val) {
            mPathRight = mRight + 1;
        }
        // compare and store longer path
        mAnswer = Math.max(mAnswer, mPathLeft + mPathRight);

        // only return one side maximum in recursive
        return Math.max(mPathLeft, mPathRight);
    }
}

File: Golang_src/medium/longestUnivaluePath_687.go
package medium

// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(h)
// Runtime: 68 ms, faster than 100.00%
// Memory Usage: 6.8 MB, less than 100.00%
func longestUnivaluePath(root *TreeNode) int {
	result := 0
	subPath(root, &result)

	return result
}

func subPath(root *TreeNode, result *int) int {
	if root == nil {
		return 0
	}
	resultL, resultR := subPath(root.Left, result), subPath(root.Right, result)

	pathL, pathR := 0, 0

	if root.Left != nil && root.Left.Val == root.Val {
		pathL = resultL + 1
	}
	if root.Right != nil && root.Right.Val == root.Val {
		pathR = resultR + 1
	}

	*result = max(*result, pathL+pathR)

	return max(pathL, pathR)
}

// Use iterative:
// Time Complexity: O(n)
// Space Complexity:O(h)
// Runtime: 108 ms, faster than 6.67%
// Memory Usage: 7.3 MB, less than 100.00%
func longestUnivaluePath2(root *TreeNode) int {
	if root == nil {
		return 0
	}

	result := 0
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		result = max(result, getLongestPathByValue(node.Left, node.Val)+getLongestPathByValue(node.Right, node.Val))

		if node.Left != nil {
			queue = append(queue, node.Left)
		}
		if node.Right != nil {
			queue = append(queue, node.Right)
		}
	}
	return result
}

func getLongestPathByValue(root *TreeNode, value int) int {
	maxDepth := 0
	if root == nil {
		return 0
	}
	queue := []*TreeNode{root}
	depth := []int{1}

	for len(queue) > 0 && len(depth) > 0 {
		node := queue[0]
		queue = queue[1:]
		pollDepth := depth[0]
		depth = depth[1:]

		if node.Val == value {
			maxDepth = max(maxDepth, pollDepth)
			if node.Left != nil {
				queue = append(queue, node.Left)
				depth = append(depth, pollDepth+1)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
				depth = append(depth, pollDepth+1)
			}
		}
	}

	return maxDepth
}
